tensorrt_llm/serve/scheduler.py

```diff
@@ -0,0 +1,217 @@
+import heapq
+from abc import ABC, abstractmethod
+from threading import Lock
+from typing import List
+
+from tensorrt_llm.serve.openai_protocol import CompletionRequest
+
+
+class Scheduler(ABC):
+
+    def __init__(self,
+                 ctx_servers: List[str] = None,
+                 gen_servers: List[str] = None):
+        self._ctx_servers = ctx_servers
+        self._gen_servers = gen_servers
+        self._lock = Lock()
+
+    def add_server(self, server: str, server_type: str):
+        raise NotImplementedError(
+            "Currently not support dynamic adding of servers")
+
+    def remove_server(self, server: str, server_type: str):
+        raise NotImplementedError(
+            "Currently not support dynamic removing of servers")
+
+    @abstractmethod
+    def schedule(self, request: CompletionRequest, server_type: str):
+        pass
+
+
+class RoundRobinScheduler(Scheduler):
+
+    def __init__(self,
+                 ctx_servers: List[str] = None,
+                 gen_servers: List[str] = None):
+        super().__init__(ctx_servers, gen_servers)
+        self._ctx_server_idx = 0
+        self._gen_server_idx = 0
+
+    def schedule(self, request: CompletionRequest, server_type: str):
+        if server_type == "context":
+            server = self._ctx_servers[self._ctx_server_idx]
+            self._ctx_server_idx = (self._ctx_server_idx + 1) % len(
+                self._ctx_servers)
+        else:
+            server = self._gen_servers[self._gen_server_idx]
+            self._gen_server_idx = (self._gen_server_idx + 1) % len(
+                self._gen_servers)
+        return server
+
+
+class LoadBalancingScheduler(Scheduler):
```

Comment:
Right now, load balanding is done using number of tokens. Should we have an option to load balance based on number of active requests?